File: app/src/main/java/com/example/words/WordsFragment.java
package com.example.words;

import android.app.AlertDialog;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.SearchView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.snackbar.Snackbar;

import java.util.List;

/**
 * A simple {@link Fragment} subclass.
 */
public class WordsFragment extends Fragment {
    private WordViewModel wordViewModel;
    private RecyclerView recyclerView;
    private MyAdapter myAdapter1, myAdapter2;
    private List<Word> allWords;
    private LiveData<List<Word>> filteredWords;
    private static final String VIEW_TYPE_SHP = "view_type_shp";
    private static final String IS_USING_CARD_VIEW = "is_using_card_view";
    private boolean undoAction;
    private DividerItemDecoration dividerItemDecoration;

    public WordsFragment() {
        // Required empty public constructor
        setHasOptionsMenu(true);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.clearData) {
            AlertDialog.Builder builder = new AlertDialog.Builder(requireActivity());
            builder.setTitle("清空数据");
            builder.setPositiveButton("确定", (dialog, which) -> wordViewModel.deleteAllWords());
            builder.setNegativeButton("取消", (dialog, which) -> {

            });
            builder.create();
            builder.show();
        } else if (item.getItemId() == R.id.switchViewType) {
            SharedPreferences sharedPreferences = requireActivity().getSharedPreferences(VIEW_TYPE_SHP, Context.MODE_PRIVATE);
            boolean viewType = sharedPreferences.getBoolean(IS_USING_CARD_VIEW, false);
            SharedPreferences.Editor editor = sharedPreferences.edit();
            if (viewType) {
                recyclerView.setAdapter(myAdapter1);
                recyclerView.addItemDecoration(dividerItemDecoration);
                editor.putBoolean(IS_USING_CARD_VIEW, false);
            } else {
                recyclerView.setAdapter(myAdapter2);
                recyclerView.removeItemDecoration(dividerItemDecoration);
                editor.putBoolean(IS_USING_CARD_VIEW, true);
            }
            editor.apply();
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.main_menu, menu);
        SearchView searchView = (SearchView) menu.findItem(R.id.app_bar_search).getActionView();
        searchView.setMaxWidth(Integer.MAX_VALUE);
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                String patten = newText.trim();
                filteredWords.removeObservers(getViewLifecycleOwner());
                filteredWords = wordViewModel.findWordsWithPatten(patten);
                filteredWords.observe(getViewLifecycleOwner(), words -> {
                    int temp = myAdapter1.getItemCount();
                    allWords = words;
                    if (temp != words.size()) {
                        myAdapter1.submitList(words);
                        myAdapter2.submitList(words);
                    }
                });
                return true;
            }
        });
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_words, container, false);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        wordViewModel = new ViewModelProvider(this).get(WordViewModel.class);
        recyclerView = requireActivity().findViewById(R.id.recyclerView);
        recyclerView.setLayoutManager(new LinearLayoutManager(requireActivity()));
        recyclerView.setItemAnimator(new DefaultItemAnimator() {
            @Override
            public void onAnimationFinished(@NonNull RecyclerView.ViewHolder viewHolder) {
                super.onAnimationFinished(viewHolder);
                LinearLayoutManager linearLayoutManager = (LinearLayoutManager) recyclerView.getLayoutManager();
                if (linearLayoutManager != null) {
                    int firstPosition = linearLayoutManager.findFirstVisibleItemPosition();
                    int lastPosition = linearLayoutManager.findLastVisibleItemPosition();
                    for (int i = firstPosition; i <= lastPosition; i++) {
                        MyAdapter.MyViewHolder holder = (MyAdapter.MyViewHolder) recyclerView.findViewHolderForAdapterPosition(i);
                        if (holder != null) {
                            holder.textViewNumber.setText(String.valueOf(i + 1));
                        }
                    }
                }
            }
        });
        myAdapter1 = new MyAdapter(false, wordViewModel);
        myAdapter2 = new MyAdapter(true, wordViewModel);

        SharedPreferences sharedPreferences = requireActivity().getSharedPreferences(VIEW_TYPE_SHP, Context.MODE_PRIVATE);
        boolean viewType = sharedPreferences.getBoolean(IS_USING_CARD_VIEW, false);
        dividerItemDecoration = new DividerItemDecoration(requireActivity(), DividerItemDecoration.VERTICAL);

        if (viewType) {
            recyclerView.setAdapter(myAdapter2);
        } else {
            recyclerView.setAdapter(myAdapter1);
            recyclerView.addItemDecoration(dividerItemDecoration);
        }

        filteredWords = wordViewModel.getAllWordsLive();
        filteredWords.observe(getViewLifecycleOwner(), words -> {
            int temp = myAdapter1.getItemCount();
            allWords = words;
            if (temp != words.size() && !undoAction) {
                if (temp < words.size()) {
                    recyclerView.smoothScrollBy(0, -200);
                }
                undoAction = false;
                myAdapter1.submitList(words);
                myAdapter2.submitList(words);
            }
        });

        new ItemTouchHelper(new ItemTouchHelper.SimpleCallback(ItemTouchHelper.UP | ItemTouchHelper.DOWN, ItemTouchHelper.START | ItemTouchHelper.END) {
            @Override
            public boolean onMove(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder, @NonNull RecyclerView.ViewHolder target) {
                Word wordFrom = allWords.get(viewHolder.getAdapterPosition());
                Word wordTo = allWords.get(target.getAdapterPosition());
                int idTemp = wordFrom.getId();
                wordFrom.setId(wordTo.getId());
                wordTo.setId(idTemp);
                wordViewModel.updateWords(wordFrom, wordTo);
                myAdapter1.notifyItemMoved(viewHolder.getAdapterPosition(), target.getAdapterPosition());
                myAdapter2.notifyItemMoved(viewHolder.getAdapterPosition(), target.getAdapterPosition());
                return false;
            }

            @Override
            public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int direction) {
                Word wordToDelete = allWords.get(viewHolder.getAdapterPosition());
                wordViewModel.deleteWords(wordToDelete);
                Snackbar.make(requireActivity().findViewById(R.id.wordsFragmentView), "删除了一个词汇", Snackbar.LENGTH_SHORT)
                        .setAction("撤销", v -> {
                            undoAction = true;
                            wordViewModel.insertWords(wordToDelete);
                        }).show();
            }
        }).attachToRecyclerView(recyclerView);
        FloatingActionButton floatingActionButton = requireActivity().findViewById(R.id.floatingActionButton);
        floatingActionButton.setOnClickListener(v -> {
            NavController navController = Navigation.findNavController(v);
            navController.navigate(R.id.action_wordsFragment_to_addFragment);
        });
    }

    @Override
    public void onResume() {
        super.onResume();
        InputMethodManager inputMethodManager = (InputMethodManager) requireActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
        inputMethodManager.hideSoftInputFromWindow(requireView().getWindowToken(), 0);
    }
}

File: app/src/main/java/com/example/words/WordsFragment.kt
package com.example.words

import android.content.Context
import android.os.Bundle
import android.view.*
import androidx.appcompat.widget.SearchView
import androidx.core.view.MenuProvider
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.booleanPreferencesKey
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.preferencesDataStore
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LiveData
import androidx.lifecycle.lifecycleScope
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.*
import com.example.words.databinding.FragmentWordsBinding
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.snackbar.Snackbar
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch

const val VIEW_TYPE_SHP = "VIEW_TYPE_SHP"
const val IS_USING_CARD_VIEW = "IS_USING_CARD_VIEW"
val Context.dataStore: DataStore<Preferences> by preferencesDataStore(VIEW_TYPE_SHP)

class WordsFragment : Fragment() {
    private var _binding: FragmentWordsBinding? = null
    private val binding get() = _binding!!
    private val viewModel by activityViewModels<WordViewModel>()
    private val dataStore by lazy { requireContext().dataStore }
    private val viewType by lazy {
        dataStore.data.map {
            it[booleanPreferencesKey(IS_USING_CARD_VIEW)] ?: false
        }.flowOn(Dispatchers.IO)
    }
    private val myAdapter1 by lazy { MyAdapter(false, viewModel) }
    private val myAdapter2 by lazy { MyAdapter(true, viewModel) }
    private val dividerItemDecoration by lazy {
        DividerItemDecoration(requireContext(), DividerItemDecoration.VERTICAL)
    }
    private val allWord by lazy { mutableListOf<Word>() }
    private lateinit var filteredWords: LiveData<List<Word>>

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentWordsBinding.inflate(inflater, container, false)
        requireActivity().addMenuProvider(
            object : MenuProvider {
                override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
                    menuInflater.inflate(R.menu.menu_main, menu)
                    val searchView: SearchView = menu.findItem(R.id.app_bar_search).actionView as SearchView
                    searchView.maxWidth = Int.MAX_VALUE
                    searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
                        override fun onQueryTextSubmit(query: String): Boolean {
                            return false
                        }

                        override fun onQueryTextChange(newText: String): Boolean {
                            newText.trim().let { pattern ->
                                filteredWords.removeObservers(viewLifecycleOwner)
                                filteredWords = viewModel.findWordsWithPattern(pattern)
                                filteredWords.observe(viewLifecycleOwner) {
                                    allWord.clear()
                                    allWord.addAll(it)
                                    if (myAdapter1.itemCount != it.size) {
                                        myAdapter1.submitList(it)
                                        myAdapter2.submitList(it)
                                    }
                                }
                            }
                            return true
                        }
                    })
                }

                override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
                    when (menuItem.itemId) {
                        R.id.app_bar_clear -> MaterialAlertDialogBuilder(requireContext())
                            .setTitle("清空数据")
                            .setPositiveButton("确定") { _, _ ->
                                viewModel.deleteAllWords()
                            }
                            .setNegativeButton("取消") { _, _ ->
                            }
                            .show()

                        R.id.app_bar_switch -> viewLifecycleOwner.lifecycleScope.launch {
                            binding.recyclerView.apply {
                                if (viewType.first()) {
                                    adapter = myAdapter1
                                    addItemDecoration(dividerItemDecoration)

                                } else {
                                    adapter = myAdapter2
                                    removeItemDecoration(dividerItemDecoration)
                                }
                            }
                            dataStore.edit {
                                it[booleanPreferencesKey(IS_USING_CARD_VIEW)] = !viewType.first()
                            }
                        }
                    }
                    return true
                }
            },
            viewLifecycleOwner,
            Lifecycle.State.RESUMED
        )
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.recyclerView.apply {
            viewLifecycleOwner.lifecycleScope.launch {
                if (viewType.first()) {
                    adapter = myAdapter2
                } else {
                    adapter = myAdapter1
                    addItemDecoration(dividerItemDecoration)
                }
            }
            itemAnimator = object : DefaultItemAnimator() {
                override fun onAnimationFinished(viewHolder: RecyclerView.ViewHolder) {
                    super.onAnimationFinished(viewHolder)
                    val linearLayoutManager = layoutManager as LinearLayoutManager
                    val firstPosition = linearLayoutManager.findFirstVisibleItemPosition()
                    val lastPosition = linearLayoutManager.findLastVisibleItemPosition()
                    for (i in firstPosition..lastPosition) {
                        (findViewHolderForAdapterPosition(i) as? MyAdapter.MyViewHolder)?.mTextViewNumber?.text =
                            getString(R.string.text_number, i + 1)
                    }
                }
            }
        }
        filteredWords = viewModel.allWordsLive
        filteredWords.observe(viewLifecycleOwner) {
            allWord.clear()
            allWord.addAll(it)
            val tmp = myAdapter1.itemCount
            if (tmp != it.size) {
                if (tmp < it.size) {
                    binding.recyclerView.smoothScrollBy(0, -200)
                }
                myAdapter1.submitList(it)
                myAdapter2.submitList(it)
            }
        }
        ItemTouchHelper(
            object : ItemTouchHelper.SimpleCallback(
                ItemTouchHelper.UP or ItemTouchHelper.DOWN,
                ItemTouchHelper.START or ItemTouchHelper.END
            ) {
                override fun onMove(
                    recyclerView: RecyclerView,
                    viewHolder: RecyclerView.ViewHolder,
                    target: RecyclerView.ViewHolder
                ): Boolean {
                    val wordFrom = allWord[viewHolder.adapterPosition]
                    val wordTo = allWord[target.adapterPosition]
                    wordFrom.id = wordTo.id.also {
                        wordTo.id = wordFrom.id
                    }
                    viewModel.updateWords(wordFrom, wordTo)
                    myAdapter1.notifyItemMoved(viewHolder.adapterPosition, target.adapterPosition)
                    myAdapter2.notifyItemMoved(viewHolder.adapterPosition, target.adapterPosition)
                    return true
                }

                override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                    val wordToDelete = allWord[viewHolder.adapterPosition]
                    viewModel.deleteWords(wordToDelete)
                    Snackbar.make(view, "删除了一个词汇", Snackbar.LENGTH_SHORT)
                        .setAction("撤销") {
                            viewModel.insertWords(wordToDelete)
                        }
                        .show()
                }
            }
        ).attachToRecyclerView(binding.recyclerView)
        binding.fab.setOnClickListener {
            findNavController().navigate(R.id.action_WordsFragment_to_AddFragment)
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding.recyclerView.adapter = null
        _binding = null
    }
}